code/modules/surgery/organs/liver.dm

```diff
@@ -12,38 +11,40 @@
 	maxHealth = STANDARD_ORGAN_THRESHOLD
 	healing_factor = STANDARD_ORGAN_HEALING
 	decay_factor = STANDARD_ORGAN_DECAY // smack in the middle of decay times
-	var/operated = FALSE	//whether we can still have our damages fixed through surgery
-	var/alcohol_tolerance = ALCOHOL_RATE//affects how much damage the liver takes from alcohol
-	var/toxTolerance = LIVER_DEFAULT_TOX_TOLERANCE//maximum amount of toxins the liver can just shrug off
-	var/toxLethality = LIVER_DEFAULT_TOX_LETHALITY//affects how much damage toxins do to the liver
-	var/filterToxins = TRUE //whether to filter toxins
-
-#define HAS_SILENT_TOXIN 0 //don't provide a feedback message if this is the only toxin present
-#define HAS_NO_TOXIN 1
-#define HAS_PAINFUL_TOXIN 2
+	/// whether we can still have our damages fixed through surgery
+	var/operated = FALSE	
+	/// affects how much damage the liver takes from alcohol
+	var/alcohol_tolerance = ALCOHOL_RATE
+	/// modifier to toxpwr for liver and toxin damage taken
+	var/toxLethality = LIVER_DEFAULT_TOX_LETHALITY 
 
 /obj/item/organ/liver/on_life()
 	var/mob/living/carbon/C = owner
 	..()	//perform general on_life()
 	if(istype(C))
 		if(!(organ_flags & ORGAN_FAILING))//can't process reagents with a failing liver
 
-			var/provide_pain_message = HAS_NO_TOXIN
-			if(filterToxins && !HAS_TRAIT(owner, TRAIT_TOXINLOVER))
-				//handle liver toxin filtration
-				for(var/datum/reagent/toxin/T in C.reagents.reagent_list)
-					var/thisamount = C.reagents.get_reagent_amount(T.type)
-					if (thisamount && thisamount <= toxTolerance)
-						C.reagents.remove_reagent(T.type, 1)
-					else
-						damage += (thisamount*toxLethality)
-						if(provide_pain_message != HAS_PAINFUL_TOXIN)
-							provide_pain_message = T.silent_toxin ? HAS_SILENT_TOXIN : HAS_PAINFUL_TOXIN
+			var/provide_pain_message = FALSE
+			var/damage_multiplier = 1 //reduced for every toxin
+			var/list/toxpwr = list()
+
+			for(var/datum/reagent/toxin/T in C.reagents.reagent_list)
+				toxpwr += T.toxpwr
+				provide_pain_message = provide_pain_message || !T.silent_toxin
+					
+			sort_list(toxpwr,  cmp = /proc/cmp_numeric_dsc) //sort all toxpowers scaling from highest to lowest
```

Comment:
I feel like this should probably be cached on the liver and updated whenever a reagent is added or removed from the body so you don't have to sort the list every single tick.

Reply:
guh, i considered that but it sounds like a lot of work